### ISSUE TITLE
Authorization and accountability

### ISSUE THREAD

**Issue by judofyr** [realistforeningen/rf-members#2]
A few issues:
- We probably want several user levels (e.g. not everyone can add lifetime members or delete old members).
- But we don't want each user of the app to have a separate account. What's going to happen is that the password to one account will to shared anyway, so we can't really protect against this.

I'm suggesting this:
- We add a `added_by (string)` column to `members` (good idea @velfundert!)
- We have two passwords that are hardcoded in the config file. One for admin, one for "regular" access.
- When users log in they type their own name and the password. The name is only for the `added_by` and it can be anything. It's not checked or anything.

**Comment by velfundert** [comment]
I suggest that we create a second table in the database with usernames, accesslevel and (hashed, smashed and salted) passwords, and a utility that generate random usernames and passwords, and automatically enter them in the database. Most of this is probably already available from whatever authorization library we are using.

This way we can use the username in `added_by`, and create and delete accounts to make sure that we always have more or less a good idea about who is adding members.

**Comment by judofyr** [comment]
Seriously, we all know what's going to happen:
- A new volunteer: "Hey, how do I use the membership app?"
- SM: "Oh, just use username=foo and password=bar"

And thus we now have _none_ accountability because everyone is using the same user.

What's never going to happen:
- Volunteers will remember their username and/or password from time to time
- SMs/KM will create new accounts to new volunteers

**Comment by velfundert** [comment]
Yes, and no: I don't think _anyone_ is going to remeber their passwords anyway, so it's better to just let them be so cheap and plentiful that you could print a couple of hundred of them, and hand out a little slip of paper with a username and password each time someone needs to log in, and of course revoke access again pretty quickly afterwards.

Think wifi-access at a hotel, not facebook-account.

**Comment by judofyr** [comment]
And how will that help? How do we map the user to a volunteer?

Nonetheless, the passwords will either (1) be locked away some place and it will annoying to fetch them or (2) they will be easy to steal.

**Comment by velfundert** [comment]
The temporary accounts could e.g. have an automatic expiration after the first time they are used. I don't think it's necessary to map the accounts to each volunteer, but it will give a certain degree of accountability. You could put the passwords the same place we kept the membership-book (the safe?). I'm absolutely convinced that it's a safer solution than to use a hardcoded password.

Anyway, those were my two cents. What we'll actually do is up to you.

**Comment by judofyr** [comment]
Well, let's back up and see what problems we're really trying to solve here. There's two problems:
1. How do we avoid bad guys to add members in the membership database
2. How do we find out who added a new member

2 is by far the most important thing. If someone finds out a new lifetime-member has been added to the database we need to be able to contact the person to verify that it's correct. If we find that e.g. all the names are lowercase we want to find the volunteer and kindly remind them that it's polite to write names in proper case. 2 is not about catching bad guys. It's about having a Real Person behind each sold membership.

I'd argue that 1 isn't that much of an issue. We will have backups of the database. We will have a system where it's impossible for anyone to destroy the data. At the end of a Friday pub they will make sure the `sum(price)` matches what they have in cash. Anything that's not registered at a Friday pub could be highlighted and verified. Added lifetime-members should be verified as well.

Actually, now that I'm writing this, I think that the most important thing about 1 is to be able to _approve/reject_ changes in the membership data, not try to avoid them happening in the first place.

What if we have a separate login before you start adding memberships?

```
Enter your name: [Bob]
Enter description of your change: [Friday pub]
```

And then an admin can list all recent changes and reject anything that seems fishy.

EDIT: So, no password required at all.

**Comment by eigilhs** [comment]
I agree with @judofyr. A password and a name should be enough. The password shouldn't be known to the volunteers, though. The SM can input this at the start of the day, and the registering page only has an editable field for the volunteer's name.

It would be cool if you had a new session when entering the password, and that the session could be accessed from more than one computer at the same time.

EDIT: I don't think registering members should be open to anyone one the web, even if new members have to be verified. It could (worst case) mean a lot of extra work at the end of the shift to figure out what happened.

**Comment by judofyr** [comment]
Sure, we can have a password as well. But we shouldn't depend on it being "secret". If someone "steals" it they should not be able to fuck things up.

**Comment by velfundert** [comment]
How about we use the IP-address of the client in the `added_by` field, don't do any authentication at all, and make sure that the server is only running when we are actually selling memberships?

**Comment by benedicteb** [comment]
I don't agree with that. Application should always be running. Maybe HS wants check member numbers++.

Shouldn't authentication be role-based? For example we create 3 accounts; one for "funker" which can only add members or edit fields. One for "styrer" which can remove, statistics ++ and a third for us, where we can do anything. Password for "funk"-account doesn't need to be that secret.

**Comment by velfundert** [comment]
I think:
1. It won't be a problem for HS to temporarily run the server if they wanted to check the memberships
2. If you want to read info about the memberships you should use the verified database, which won't be the same as the one this app is using.
3. If you want proper statistics you probably want to use a different app anyway 
4. If we want anyone to be able to check their memberships, we should make a different app that doesn't need authorization, and reads from the verified database.

**Comment by judofyr** [comment]
There's no reason to shut _down_ the server. But yes, we can make it possible to only add members in certain time periods. Not exactly sure why we'd like to do that? It doesn't really make it any more secure; it just means that any malice will be done simultaneously with a Friday pub.

Storing IP-addresses are a good idea though. Then we can show whether a request comes from inside the UiO network or outside.

Again, coupling the adding of members to a "session" seems like a good idea: When you want to add new members you must first start a new session:

```
Your name: [Bob]
Description: [Friday pub]
Password: [A not very secret password]

[Start adding members!]
```

You can always edit memberships _within_ your session, but the moment you finalize the session (click a "I'm done!"-button), it's locked.

We'll have a separate list that shows all sessions with the number of memberships added:

```
For SM/KM only:

| Name   | Description | Memberships | Price    |
+--------+-------------+-------------+----------+
| Magnus | Friday pub  | 57          | 2.850,00 |
```

There could also be a button for marking a full session for invalid, or listing all memberships added in a session. There should probably be a way to mark a session as "accepted" which is done after SM has verified the amount in the membership database is in sync with the money at hand.

**Comment by velfundert** [comment]
On friday, @skjvlnd wrote:

> A password and a name should be enough. The password shouldn't be known to the volunteers, though. The SM can input this at the start of the day, and the registering page only has an editable field for the volunteer's name.
> 
> It would be cool if you had a new session when entering the password, and that the session could be accessed from more than one computer at the same time.

The easiest way to do this is to simply not have the server running other than when a SM actually wants to be able to add membeships.

**Comment by judofyr** [comment]
> The easiest way to do this is to simply not have the server running other than when a SM actually wants to be able to add membeships.

Again: It doesn't protect against anything. Anyone who wants to do malice will be able to do it Mon-Fri 10-16 and all Friday night.

**Comment by judofyr** [comment]
I also don't think malice will be such a big problem. That's why I want to do as little as possible initially. Make it possible for anyone to add members. Make it really easy and user friendly. 

As long as we can track anything that has been added, we can clean it up. If it becomes a real problem we can add more protection/hassle later.

**Comment by eigilhs** [comment]
Simple password protection also ensures that no crawlers (or anyone else) get a hold of our member list. I think the simplest would be for the SM to log in, and have a cookie lasting for the entire shift. That way it's not necessary for the volunteers to log in and out ten times in a night, they just have to make sure that it's their name that's in the "added by" box.

**Comment by benedicteb** [comment]
We're discussing this now. When we go into production there should at least be some form of protection. No log-in or only give your name at log-in isn't a good enough precaution considering we store full names of all members.

Temporarily a single user everyone uses at least?

**Comment by judofyr** [comment]
I've implemented this in #13:
- There's four levels: Funk, SM, Admin, Superadmin. You need to login before you can do anything at all. When you login we create a session (which contains the current time and a description).
- Everything you do (creating memberships) are connected to a specific session.
- We can then filter memberships by the sessions they were created in.
- You can only delete memberships that was created in your own session.

This means that if someone manages to steal a password, we will be able to see all the changes they've done.
